src/CustomMarshalers/src/System/Runtime/InteropServices/CustomMarshalers/ComDataHelpers.cs

```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace System.Runtime.InteropServices.CustomMarshalers
+{
+    internal static class ComDataHelpers
+    {
+        public static TView GetOrCreateManagedViewFromComData<T, TView>(object comObject, Func<T, TView> createCallback)
+        {
+            object key = typeof(TView);
+
+            if (Marshal.GetComObjectData(comObject, key) is TView managedView)
+            {
+                return managedView;
+            }
+            else
+            {
+                managedView = createCallback((T)comObject);
+                if (!Marshal.SetComObjectData(comObject, key, managedView))
+                {
+                    managedView = (TView)Marshal.GetComObjectData(comObject, key);
```

Comment:
Is it possible ComObjectData could have been removed for this object/key?  If so, could it have been removed between the Set above failing and this Get call?

Reply:
In the general case that is possible. However, the keys used for the data are all `System.Type` objects representing internal types. The only possible way I see someone clearing it without private reflection would be something like the following:

- The user uses one of these marshalers to get the managed object.
- They call `GetType()` on that object. This gets them the key.
- In another call, they have the COM object be marshalled without using the custom marshaler, so they get the RCW.
- They call `Marshal.SetComObjectData` with the `Type` object from above and the RCW to remove the data.

Reply:
This same pattern is used in the `System.__ComObject` type directly: https://github.com/dotnet/coreclr/blob/87ac2b53d4618f994e67266fd99f479531018059/src/System.Private.CoreLib/src/System/__ComObject.cs#L147-166